dcicutils/license_utils.py

```diff
@@ -0,0 +1,806 @@
+                PRINT(f" {license}: {', '.join(names)}")
+        return analysis
+
+    @classmethod
+    def validate(cls, frameworks: Optional[List[FrameworkSpec]] = None) -> None:
+        """
+        This method is intended to be used in a unit test, as in:
+
+            from my_org_tools import MyOrgLicenseChecker
+            def test_license_compatibility():
+                MyOrgLicenseChecker.validate()
+
+        where my_org_tools has done something like:
+
+            from dcicutils.license_utils import LicenseChecker
+            class MyOrgLicenseChecker(LicenseChecker):
+                LICENSE_OWNER = "..."
+                ALLOWED = [...]
+                EXPECTED_MISSING_LICENSES = [...]
+                EXCEPTIONS = {...}
+
+        See the example of C4InfrastructureLicenseChecker we use in our own group for our own family of toools,
+        which we sometimes informally refer to collectively as 'C4'.
+        """
+        analysis = LicenseAnalysis()
+        cls.analyze_license_dependencies_by_framework(analysis=analysis, frameworks=frameworks)
+        cls.analyze_license_file(analysis=analysis)
+        cls.show_unacceptable_licenses(analysis=analysis)
+        if analysis.unexpected_missing:
+            logger.warning(there_are(analysis.unexpected_missing, kind='unexpectedly missing license', punctuate=True))
+        if analysis.no_longer_missing:
+            logger.warning(there_are(analysis.no_longer_missing, kind='no-longer-missing license', punctuate=True))
+        for message in analysis.miscellaneous:
+            logger.warning(message)
+        if analysis.unacceptable:
+            raise LicenseAcceptabilityCheckFailure(unacceptable_licenses=analysis.unacceptable)
+
+
+class LicenseCheckFailure(Exception):
+
+    DEFAULT_MESSAGE = "License check failure."
+
+    def __init__(self, message=None):
+        super().__init__(message or self.DEFAULT_MESSAGE)
+
+
+class LicenseOwnershipCheckFailure(LicenseCheckFailure):
+
+    DEFAULT_MESSAGE = "License ownership check failure."
+
+
+class LicenseAcceptabilityCheckFailure(LicenseCheckFailure):
+
+    DEFAULT_MESSAGE = "License acceptability check failure."
+
+    def __init__(self, message=None, unacceptable_licenses=None):
+        self.unacceptable_licenses = unacceptable_licenses
+        if not message and unacceptable_licenses:
+            message = there_are(unacceptable_licenses, kind='unacceptable license')
+        super().__init__(message=message)
+
+
+class C4InfrastructureLicenseChecker(LicenseChecker):
+    """
+    This set of values is useful to us in Park Lab where these tools were developed.
+    If you're at some other organization, we recommend you make a class that has values
+    suitable to your own organizational needs.
+    """
+
+    COPYRIGHT_OWNER = "President and Fellows of Harvard College"
+    LICENSE_TITLE = "The MIT License"
+    LICENSE_FRAMEWORKS = ['python', 'javascript']
+
+    ALLOWED = [
```

Comment:
For our purposes it may be useful to be able to read this from file, so we can write top level files with this information or even a directory that can have sane defaults in utils and be overridden in other repos

Reply:
As a matter of design, I did not want individual repos to decide to override it. We can have several named policies. But the policies need to be approved organizationally, not at the level of coding lines.  I will likely add one for pipelines.

I have tried this in a number of repos (see the list on the first page). Some problems turned up that were since fixed. Others are ongoing issues for which I may create a way to mark the license as a known problem. We can talk more about this interactively perhaps sometime.